storage/innobase/xtrabackup/src/ds_decompress.cc

```diff
@@ -0,0 +1,600 @@
+        if (decomp_file->decomp_ctxt->chunk_size == 0) {
+          decomp_file->decomp_ctxt->chunk_size = chunk_size;
+          initialize_worker_thread_buffers(
+            decomp_file->decomp_ctxt->threads,
+            decomp_file->decomp_ctxt->nthreads,
+            chunk_size);
+        } else if (chunk_size != decomp_file->decomp_ctxt->chunk_size) {
+          /* all files in single archive should be using the same chunk size */
+          msg("decompress: multiple chunk sizes found: %lld != %lld.\n",
+              chunk_size, decomp_file->decomp_ctxt->chunk_size);
+          return 1;
+        }
+        decomp_file->state = STATE_READ_FILE_HEADER;
+        decomp_file->nbytes_expected = 5;
+      } else if (decomp_file->state == STATE_READ_FILE_HEADER) {
+        if (header[0] != 'F') {
+          msg("decompress: invalid file header, 'F' not found.\n");
+          return 1;
+        }
+        decomp_file->state = STATE_SKIP_FILE_NAME;
+        /* + 1 because file name length does not include zero terminator */
+        decomp_file->nbytes_expected = uint4korr(header + 1) + 1;
+      } else if (decomp_file->state == STATE_SKIP_FILE_NAME) {
+        decomp_file->state = STATE_DETECT_BLOCK_OR_EOF;
+        decomp_file->nbytes_expected = 1;
+      } else if (decomp_file->state == STATE_DETECT_BLOCK_OR_EOF) {
+        if (header[0] == 'N') {
+          decomp_file->state = STATE_READ_BLOCK_HEADER;
+          decomp_file->nbytes_expected = 19;
+        } else if (header[0] == 'E') {
+          decomp_file->state = STATE_READ_FILE_TRAILER;
+          decomp_file->nbytes_expected = 15;
+        } else {
+          msg("decompress: invalid block header, 'N' or 'E' not found.\n");
+          return 1;
+        }
+      } else if (decomp_file->state == STATE_READ_BLOCK_HEADER) {
+        if (strncmp(header, "EWBNEWB", 7)) {
+          msg("decompress: invalid block header, 'NEWBNEWB' not found.\n");
+          return 1;
+        }
+        decomp_file->state = STATE_READ_BLOCK_SIZE_HEADER;
+        decomp_file->nbytes_expected = 9;
+      } else if (decomp_file->state == STATE_READ_FILE_TRAILER) {
+        if (strncmp(header, "NDSENDS", 7)) {
+          msg("decompress: invalid file trailing, 'ENDSENDS' not found.\n");
+          return 1;
+        }
+        decomp_file->state = STATE_REACHED_EOF;
+      }
+    } else if (decomp_file->state == STATE_READ_BLOCK_SIZE_HEADER) {
+      /* Block size handling requires a bit of extra effort because the
+      decompress code expects size bytes to be part of the data buffer */
+      if (decomp_file->nbytes_read == 0 &&
+          nbytes_available == nbytes_missing) {
+        decomp_file->nbytes_expected = qlz_size_compressed(*ptr);
+        if ((*len) >= decomp_file->nbytes_expected) {
+          decomp_file->state = STATE_PROCESS_DIRECT_BLOCK_DATA;
+          return 0;
+        } else {
+          decomp_file->state = STATE_BUFFER_BLOCK_DATA;
+          decomp_file->buffer = (char *)my_malloc(
```

Comment:
It looks like `decomp_file->buffer` leaks.

Reply:
It shouldn't. This code is ever reached after previously allocated buffer has been moved to `decomp_thread_ctxt_t` of one of the threads and that buffer is freed separately after the thread finishes processing the data.

Reply:
I see. I missed that bit.